File: conversion_tool.py
class Converter:
    def binary_to_decimal(self, binary_num):
        return int(binary_num, 2)

    def octal_to_decimal(self, octal_num):
        return int(octal_num, 8)

    def hexadecimal_to_decimal(self, hex_num):
        return int(hex_num, 16)

    def decimal_to_binary(self, decimal_num):
        return bin(decimal_num)

    def decimal_to_octal(self, decimal_num):
        return oct(decimal_num)

    def decimal_to_hexadecimal(self, decimal_num):
        return hex(decimal_num)

    def bytes_to_bits(self, byte_num):
        return byte_num * 8

    def bits_to_bytes(self, bit_num):
        return bit_num // 8


class ConversionTool:
    def __init__(self):
        self.converter = Converter()

    def run(self):
        while True:
            print("\nChoose an option:")
            print("1. Convert Binary to Decimal")
            print("2. Convert Octal to Decimal")
            print("3. Convert Hexadecimal to Decimal")
            print("4. Convert Decimal to Binary")
            print("5. Convert Decimal to Octal")
            print("6. Convert Decimal to Hexadecimal")
            print("7. Convert Bytes to Bits")
            print("8. Convert Bits to Bytes")
            print("9. Quit")

            choice = input("Enter your choice (1/2/3/4/5/6/7/8/9): ")

            match choice:
                case '1':
                    binary_num = input("Enter a binary number: ")
                    result = self.converter.binary_to_decimal(binary_num)
                    print(f"The decimal equivalent of {binary_num} is {result}")

                case '2':
                    octal_num = input("Enter an octal number: ")
                    result = self.converter.octal_to_decimal(octal_num)
                    print(f"The decimal equivalent of {octal_num} is {result}")

                case '3':
                    hex_num = input("Enter a hexadecimal number: ")
                    result = self.converter.hexadecimal_to_decimal(hex_num)
                    print(f"The decimal equivalent of {hex_num} is {result}")

                case '4':
                    decimal_num = int(input("Enter a decimal number: "))
                    result = self.converter.decimal_to_binary(decimal_num)
                    print(f"The binary equivalent of {decimal_num} is {result}")

                case '5':
                    decimal_num = int(input("Enter a decimal number: "))
                    result = self.converter.decimal_to_octal(decimal_num)
                    print(f"The octal equivalent of {decimal_num} is {result}")

                case '6':
                    decimal_num = int(input("Enter a decimal number: "))
                    result = self.converter.decimal_to_hexadecimal(decimal_num)
                    print(f"The hexadecimal equivalent of {decimal_num} is {result}")

                case '7':
                    byte_num = int(input("Enter the number of bytes: "))
                    result = self.converter.bytes_to_bits(byte_num)
                    print(f"{byte_num} bytes are equal to {result} bits.")

                case '8':
                    bit_num = int(input("Enter the number of bits: "))
                    result = self.converter.bits_to_bytes(bit_num)
                    print(f"{bit_num} bits are equal to {result} bytes.")

                case '9':
                    print("Thank you for using the conversion program! Goodbye!")
                    return

                case _:
                    print("Invalid choice. Please choose a valid option.")


if __name__ == "__main__":
    conversion_tool = ConversionTool()
    conversion_tool.run()

File: conversion_tool.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <iomanip>

class Converter {
public:
    long long binaryToDecimal(const std::string& binaryNum) {
        return std::stoll(binaryNum, 0, 2);
    }

    long long octalToDecimal(const std::string& octalNum) {
        return std::stoll(octalNum, 0, 8);
    }

    long long hexadecimalToDecimal(const std::string& hexNum) {
        return std::stoll(hexNum, 0, 16);
    }

    std::string decimalToBinary(long long decimalNum) {
        std::stringstream ss;
        ss << std::bitset<64>(decimalNum);
        return ss.str();
    }

    std::string decimalToOctal(long long decimalNum) {
        std::stringstream ss;
        ss << std::oct << decimalNum;
        return ss.str();
    }

    std::string decimalToHexadecimal(long long decimalNum) {
        std::stringstream ss;
        ss << std::hex << decimalNum;
        return ss.str();
    }

    long long bytesToBits(long long byteNum) {
        return byteNum * 8;
    }

    long long bitsToBytes(long long bitNum) {
        return bitNum / 8;
    }
};

class ConversionTool {
public:
    ConversionTool() {
        converter = Converter();
    }

    void run() {
        while (true) {
            std::cout << "\nChoose an option:" << std::endl;
            std::cout << "1. Convert Binary to Decimal" << std::endl;
            std::cout << "2. Convert Octal to Decimal" << std::endl;
            std::cout << "3. Convert Hexadecimal to Decimal" << std::endl;
            std::cout << "4. Convert Decimal to Binary" << std::endl;
            std::cout << "5. Convert Decimal to Octal" << std::endl;
            std::cout << "6. Convert Decimal to Hexadecimal" << std::endl;
            std::cout << "7. Convert Bytes to Bits" << std::endl;
            std::cout << "8. Convert Bits to Bytes" << std::endl;
            std::cout << "9. Quit" << std::endl;

            int choice;
            std::cin >> choice;

            switch (choice) {
                case 1: {
                    std::string binaryNum;
                    std::cout << "Enter a binary number: ";
                    std::cin >> binaryNum;
                    long long result = converter.binaryToDecimal(binaryNum);
                    std::cout << "The decimal equivalent is " << result << std::endl;
                    break;
                }

                case 2: {
                    std::string octalNum;
                    std::cout << "Enter an octal number: ";
                    std::cin >> octalNum;
                    long long result = converter.octalToDecimal(octalNum);
                    std::cout << "The decimal equivalent is " << result << std::endl;
                    break;
                }

                case 3: {
                    std::string hexNum;
                    std::cout << "Enter a hexadecimal number: ";
                    std::cin >> hexNum;
                    long long result = converter.hexadecimalToDecimal(hexNum);
                    std::cout << "The decimal equivalent is " << result << std::endl;
                    break;
                }

                case 4: {
                    long long decimalNum;
                    std::cout << "Enter a decimal number: ";
                    std::cin >> decimalNum;
                    std::string result = converter.decimalToBinary(decimalNum);
                    std::cout << "The binary equivalent is " << result << std::endl;
                    break;
                }

                case 5: {
                    long long decimalNum;
                    std::cout << "Enter a decimal number: ";
                    std::cin >> decimalNum;
                    std::string result = converter.decimalToOctal(decimalNum);
                    std::cout << "The octal equivalent is " << result << std::endl;
                    break;
                }

                case 6: {
                    long long decimalNum;
                    std::cout << "Enter a decimal number: ";
                    std::cin >> decimalNum;
                    std::string result = converter.decimalToHexadecimal(decimalNum);
                    std::cout << "The hexadecimal equivalent is " << result << std::endl;
                    break;
                }

                case 7: {
                    long long byteNum;
                    std::cout << "Enter the number of bytes: ";
                    std::cin >> byteNum;
                    long long result = converter.bytesToBits(byteNum);
                    std::cout << byteNum << " bytes are equal to " << result << " bits." << std::endl;
                    break;
                }

                case 8: {
                    long long bitNum;
                    std::cout << "Enter the number of bits: ";
                    std::cin >> bitNum;
                    long long result = converter.bitsToBytes(bitNum);
                    std::cout << bitNum << " bits are equal to " << result << " bytes." << std::endl;
                    break;
                }

                case 9:
                    std::cout << "Thank you for using the conversion program! Goodbye!" << std::endl;
                    return;

                default:
                    std::cout << "Invalid choice. Please choose a valid option." << std::endl;
                    break;
            }
        }
    }

private:
    Converter converter;
};

int main() {
    ConversionTool conversionTool;
    conversionTool.run();
    return 0;
}
